sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -579,36 +582,62 @@ static inline int find_codec_info_acpi(const u8 *acpi_id)
 	return i;
 }
 
-static int get_sdw_dai_link_num(struct snd_soc_acpi_mach_params *mach_params)
+/*
+ * get BE dailink number and CPU DAI number based on sdw link adr.
+ * Since some sdw links may be aggregated, the CPU DAI number may be
+ * larger than the number of BE dailinks.
+ */
+static int get_sdw_dailink_info(const struct snd_soc_acpi_link_adr *links,
+				int *sdw_be_num, int *sdw_cpu_dai_num)
 {
-	const struct snd_soc_acpi_link_adr *link;
-	int num = 0;
+	const struct snd_soc_acpi_link_adr *link = links;
+	bool group_visited[SDW_MAX_GROUPS];
+	int cpu_dai_num = 0;
+	int be_num = 0;
+	int i;
 
-	link = mach_params->links;
-	if (!link)
+	if (!links)
 		return -EINVAL;
 
+	for (i = 0; i < SDW_MAX_GROUPS; i++)
+		group_visited[i] = false;
+
 	for (; link->num_adr; link++) {
+		const struct snd_soc_acpi_endpoint *endpoints;
 		int part_id, id;
 		u64 adr;
 
-		adr = link->adr_d[0].adr;
+		endpoints = link->adr_d->endpoints;
+		/* for aggregated link only increment cpu_dai_num */
+		if (endpoints->aggregated &&
+		    group_visited[endpoints->group_id]) {
+			cpu_dai_num++;
+			continue;
+		}
+
+		adr = link->adr_d->adr;
 
 		part_id = SDW_PART_ID(adr);
 		id = find_codec_info_part(part_id);
 		if (id < 0)
-			return 0;
-
-		/* support playback */
-		if (codec_info_list[id].direction[0])
-			num++;
+			return -EINVAL;
+
+		for (i = SNDRV_PCM_STREAM_PLAYBACK;
+			i <= SNDRV_PCM_STREAM_CAPTURE; i++) {
+			if (codec_info_list[id].direction[i]) {
+				cpu_dai_num++;
+				be_num++;
```

Comment:
@RanderWang I'm trying to understand the logic here in get_sdw_dailink_info(). We would like to get the number of CPU DAI's and the number of BE DAI links, isnt it? The number of CPU DAI's depends on the number of endpoints on each adr on each link isnt it? So in order to count the cpu dais, we dont really need to care about the whether the endpoint is aggregated is not right?

Reply:
yes, the cpu dais is not connected to aggregated status, only BE DAI

Reply:
I guess you found a optimal way to process them, let't me have a try.

Reply:
@RanderWang does this belong inside the for_each_pcm_streams(stream)? You will count the be_num twice for non-aggregated cases isnt it? whereas if you put it outside the for_each, you can simply do (I think this is right, but please confirm):
```
if (!endpoint->group_id || !group_visited[endpoint->group_id])
        be_num++;
```